kroxylicious/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaMessageDecoder.java

```diff
@@ -25,7 +25,7 @@ public KafkaMessageDecoder() {
     }
 
     @Override
-    protected synchronized void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+    public synchronized void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
```

Comment:
as an aside: I have a hazy memory that its more performant to synchronise on an object rather than at the method level is that something we should be considering here.

Reply:
I guess you meant [this](https://dzone.com/articles/escape-analysis). I don't object, but I wonder if @franz1981 has any insights to lend here (e.g. maybe it's no necessary single the JVM would usually be able to figure this out anyway in a typical netty handler).

Reply:
I think it was mostly a case of miss applying this sentiment from effective java:
>More generally, keep the amount of work that you do from within synchronized regions to a minimum
Coupled with generally selecting something other than `this` for the synchronised lock. Having said that what is the scope of a `KafkaMessageDecoder` is it one per pipeline or one per instance, as it might be more appropriate to sychronise on the byetbuf rather than the decoder. We don't want multiple calls to the decoder contending on its lock when they are trying to process independent buffers.

Reply:
> Having said that what is the scope of a `KafkaMessageDecoder` is it one per pipeline or one per instance

Looking at the constructor call hierarchy would tell you that it's one per pipeline.

I think what you're really missing here is an answer to the question: "Why is KafkaMessageDecoder using `synchronized` here?" It's shame on me that a) it's not documented, and b) if there was a good reason it now escapes me. Looking back over the history it got added in the 3rd commit where I also removed a field, so I'm guessing I added while the member was there and later removed the member but didn't remove the `synchronized` and then committed. I'll remove it. Thank you!

Reply:
That's my question indeed :)